### PR TITLE
WIP initial work adding ctx to client-gen and informer-gen

### PR DESCRIPTION
**Pull request by maleck13**
**What this PR does / why we need it**:
Adds changes to the code generators to change the API to allow passing in context.Context as the first argument
https://github.com/kubernetes/community/pull/1166
**Special notes for your reviewer**:
pretty new to the code here so all guidance is appreciated. Fairly happy with the client gen changes and have tested those out on a CRD client I have, less certain about the Informer changes.
Wondered if perhaps trying them with the sample-controller would be a good approach.
**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by maleck13** [comment]
@sttts wondering did you get chance to take a look. Really just looking for feedback that heading in right direction with this before going to far

**Comment by sttts** [comment]
I have put this on the agenda of tomorrow's sig-api-machinery meeting https://docs.google.com/a/google.com/document/d/1x9RNaaysyO0gXHIr1y50QFbiL1x8OWnk2v3XnrdkT5Y/edit?usp=drive_web

**Comment by maleck13** [comment]
@sttts thanks. Do I need to attend that meeting?

**Comment by sttts** [comment]
> Do I need to attend that meeting?

Depends on how curious you are :)

**Comment by maleck13** [comment]
I can't seem to access that document, should I be able to?

**Comment by caesarxuchao** [comment]
cc @hongchaodeng

**Comment by lavalamp** [comment]
The sig was in favor of this change, btw.

(you can access the doc by joining the sig mailing list, or the kubernetes-dev mailing list)

/ok-to-test

**Comment by sttts** [comment]
@maleck13 anything missing we have to do for an early change shortly after master opens again?

**Comment by maleck13** [comment]
@sttts I would like to try the changes on the sample controller to ensure I have covered the require changes for informers and listers. I will try get to this by the end of the week

**Comment by sttts** [comment]
@maleck13 as master is closed right now, we are not in a hurry. But it would easier to have all this ready before master opens again.

**Comment by maleck13** [comment]
@sttts ok any timeline on when master opens again. I think a good option here would be to create PRs against the sample apiserver and the sample controller after running the new generators as it will show all the sites that have changed and give examples for developers on how to update their own projects. WDYT

**Comment by maleck13** [comment]
@sttts reasonably happy with the changes here, I used the new gen against the sample-controller and the sample-api server. All looked good. I have included the generated client etc for the sample-apiserver as a second commit here. 
What are the next steps here? Is it a case of running the generator against the main repo and updating all the call sites. Going to be a big PR :)

**Comment by maleck13** [comment]
Am I correct in thinking that the expansions such as ```pkg/client/clientset_generated/internalclientset/typed/core/internalversion/event_expansion.go``` are not actually generated?

**Comment by sttts** [comment]
@maleck13 right. The expansions are the extension point to add methods manually.

**Comment by maleck13** [comment]
Ok thanks, so I will update those manually to accept the context. Unfortunately I am struggling with time at the moment due to many other commitments. I believe the generator changes are good now so need to update all the call sites and change the extensions

**Comment by sttts** [comment]
> I believe the generator changes are good now so need to update all the call sites

This means to use the existing context from the RequestContextMapper in the apiserver and context.TODO() elsewhere? What about the ctrl managers and kubelet? Do we have contexts there already?

**Comment by maleck13** [comment]
> This means to use the existing context from the RequestContextMapper in the apiserver and context.TODO() elsewhere? 
Yes this makes sense to me

> What about the ctrl managers and kubelet? Do we have contexts there already?
I don't believe so, I will take a closer look. If there is no context available I had intended to also update these call sites with ```context.TODO()```

**Comment by sttts** [comment]
@maleck13 what is the status of this? Can we continue with this in 1.11?

/cc @dungeonmaster18 @gargankur74

**Comment by maleck13** [comment]
Yes this I still want to see this through, time has been my issue. I have scheduled time next week to get back to this

**Comment by maleck13** [comment]
@sttts Thanks. As I see I have the following tasks.

- [ ] Run the new generator to update the clients ( I have done this and it looked to work as expected, but need to rebase and do it again)
- [ ] Update all the call sites appropriately (using context.ToDO or an existing context if there is one)
- [ ] Update the extensions to the clients to accept a context

**Comment by gargankur74** [comment]
@maleck13 I will start with running the new generator to update the clients and testing whether it works as expected.

**Comment by maleck13** [comment]
The flow I am following to do this work is as follows:

1) Update to the latest master (should this work go from the latest release?)
2) Use ```hack/update-codegen.sh``` to generate all the things
3) Pick an API update the signatures, look for any extensions and change these. 

@sttts Looking for a little guidance here. Is there a better way to approach such a large change in your experience. 
Would it be a better option, for instance to generate the clients for a single api group update the call sites commit do the next one etc?

**Comment by dungeonmaster18** [comment]
Hey @maleck13, I have cloned your repo and removed the conflicts and run hack/update-codegen.sh. I just wanted to know how do you verify the generated code is working fine. I tried running hack/verify-codegen.sh but not sure if it's the right approach. Any help on that

**Comment by maleck13** [comment]
> I just wanted to know how do you verify the generated code is working fine.
I am not aware of a way of doing this other than getting the code to compile and the tests to execute. There will be many call sites within the code after running the generate where there will be no ctx passed through. Most of these will need to be updated with ```context.TODO```

**Comment by maleck13** [comment]
Not entirely sure this is the right thing to do, but I am going to push two commits one after running the code gen and a second show some call site updates. Hoping this will give people opportunity to tell me if I am approaching this incorrectly.

**Comment by maleck13** [comment]
This obviously wont compile currently

**Comment by dungeonmaster18** [comment]
Hey @maleck13, I see you are updating call sites as well. I also started with that. I think we should have a different pr for call site update because after running codegen this pr already become very large and I guess having different pr will make it easy to review. Also if you don't mind I will update the call site because I have already updated half of them. If this approach looks good then I will continue working on call site update.

**Comment by dungeonmaster18** [comment]
Hey @maleck13 I have replaced all the on site calls and all the test cases were passing. Now I wanted to to know, how should we go with this PR. 

CC/ @sttts  need your help.

**Comment by maleck13** [comment]
@dungeonmaster18 Thats awesome that you have done this work and tests looking good! Personally I would keep it as part of this PR as an individual commit that way the changes to code generator can be reviewed the output of the code generator can be reviewed in the separate commit on this branch and then the updates to the call sites can be reviewed. 
The changes shouldn't be difficult to review I think, as they are likely very much the same change over and over. I will of course defer to @sttts  on this

**Comment by dungeonmaster18** [comment]
Cool then. In case you need any help regarding call site updates just ping
me.

**Comment by maleck13** [comment]
@dungeonmaster18 I want to make use of your changes if possible rather than redo them. Perhaps I can cherry pick your changes onto this pr?

**Comment by maleck13** [comment]
@dungeonmaster18 are you able to push up your changes?

**Comment by dungeonmaster18** [comment]
Hey @maleck13, sorry for the late reply. Actually I run the test on master by mistake and later realized it. But anyway i have more than 100 files updated. Also you have to update https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/code-generator/cmd/client-gen/generators/fake/generator_fake_for_type.go this file with context in order to generate fake clients.

**Comment by maleck13** [comment]
@dungeonmaster18 good spot on the fakes will update

**Comment by dungeonmaster18** [comment]
Hey @maleck13 have you tested client-gen changes against client-go. Because there was some issue with admissionregistration package which is auto generated by client gen.

**Comment by sttts** [comment]
> they are symlinked so this question can be ignored

Yes, they are in staging/ in k/k.

**Comment by dungeonmaster18** [comment]
Hey @maleck13, I have changes all the call sites and here it is https://github.com/dungeonmaster18/kubernetes/commits/add-context-for-onsite-call . There are still few error which are due to generated files. Please have a look and update those code-generator files accordingly.

**Comment by PhilipGough** [comment]
Hey @dungeonmaster18, I had also started this work so I'm going to cherry pick your changes. Thanks

**Comment by dungeonmaster18** [comment]
Hey @PhilipGough , All the call sites are replaced but there was some problem in generated files. I didn't get time to look into that. What I probably suspect that it is due to the auto generated type interface. Please have a look and see if you can fix that.

**Comment by PhilipGough** [comment]
Have updated any files, including tests that require context. See https://github.com/maleck13/kubernetes/pull/1

**Comment by dungeonmaster18** [comment]
Does ```make test``` passes all the tests without throwing any error.

**Comment by PhilipGough** [comment]
> Does make test passes all the tests without throwing any error.

@dungeonmaster18 not as of yet. Will revisit tomorrow and attempt to fix failures

**Comment by maleck13** [comment]
Will cherry pick @PhilipGough changes. Gave them a quick scan and look reasonable. That way we can see if the tests are all passing and start the review update process

**Comment by maleck13** [comment]
needs a serious rebase will do this next

**Comment by sttts** [comment]
@maleck13 should we try to get this ready while master is frozen and then merge early on? Looks like a rebase hell otherwise.

**Comment by maleck13** [comment]
@sttts yes @PhilipGough  is looking at the rebase. When does master reopen?

**Comment by sttts** [comment]
> When does master reopen?

traditionally it takes some weeks, maybe end of June?

**Comment by PhilipGough** [comment]
@sttts, @maleck13 yes, have started the rebase. Will keep updated on progress

**Comment by ashmrtn** [comment]
Is this PR expected to land soon-ish? I would also be interested in using context with OpenTracing (at least to get proper span views by setting the parent of a span) in a project I am working on

**Comment by PhilipGough** [comment]
@sttts Hey, I had begun the rebasing of this, but then some stuff came up and I needed to park it. Only available to get back to it from today. What is the best way to proceed now since we missed the June date on this?

**Comment by sttts** [comment]
> What is the best way to proceed now since we missed the June date on this?

I think it has not changed much. The change rate of the repo is not that big right now. So, the fast we have this ready the easier it will be.

**Comment by PhilipGough** [comment]
@maleck13 @sttts I've regenerated and rebased, updated remaining call sites and  run `make test` on  https://github.com/PhilipGough/kubernetes/tree/context-dot-todo

@maleck13 can you take a look at maybe do a hard reset to my branch is probably the cleanest solution here if you are happy

**Comment by maleck13** [comment]
@PhilipGough done

**Comment by PhilipGough** [comment]
@maleck13 @sttts I've rebased off master, regenerated the clients, update any additional call sites and fixed as many tests as I can run locally. Going to need some help on fixing the remaining failures.

**Comment by sttts** [comment]
```
untime: goroutine stack exceeds 1000000000-byte limit
fatal error: stack overflow

runtime stack:
runtime.throw(0x409d7f7, 0xe)
	/usr/local/go/src/runtime/panic.go:616 +0x81
runtime.newstack()
	/usr/local/go/src/runtime/stack.go:1054 +0x71f
runtime.morestack()
	/usr/local/go/src/runtime/asm_amd64.s:480 +0x89

goroutine 2446 [running]:
k8s.io/kubernetes/pkg/scheduler/factory.(*podInformer).Informer(0xc424a77d30, 0x55b69e0, 0xc420056160, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/factory/factory.go:1280 +0x62 fp=0xc447624378 sp=0xc447624370 pc=0x11d38c2
k8s.io/kubernetes/pkg/scheduler/factory.(*podInformer).Informer(0xc424a77d30, 0x55b69e0, 0xc420056160, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/factory/factory.go:1281 +0x3f fp=0xc4476243b0 sp=0xc447624378 pc=0x11d389f
k8s.io/kubernetes/pkg/scheduler/factory.(*podInformer).Informer(0xc424a77d30, 0x55b69e0, 0xc420056160, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/factory/factory.go:1281 +0x3f fp=0xc4476243e8 sp=0xc4476243b0 pc=0x11d389f
```

**Comment by PhilipGough** [comment]
@sttts thanks

File: cmd/kube-controller-manager/app/core.go [hunk 1]
```diff
@@ -57,6 +57,7 @@ import (
 	"k8s.io/kubernetes/pkg/quota/generic"
 	quotainstall "k8s.io/kubernetes/pkg/quota/install"
 	"k8s.io/kubernetes/pkg/util/metrics"
+	"context"
```
Comment (sttts) [inline cmd/kube-controller-manager/app/core.go]: import order

File: cmd/kubeadm/app/discovery/file/file.go [hunk 2]
```diff
@@ -72,7 +73,7 @@ func ValidateClusterInfo(clusterinfo *clientcmdapi.Config, clustername string) (
 	var clusterinfoCM *v1.ConfigMap
 	wait.PollInfinite(constants.DiscoveryRetryInterval, func() (bool, error) {
 		var err error
-		clusterinfoCM, err = client.CoreV1().ConfigMaps(metav1.NamespacePublic).Get(bootstrapapi.ConfigMapClusterInfo, metav1.GetOptions{})
+		clusterinfoCM, err = client.CoreV1().ConfigMaps(metav1.NamespacePublic).Get(context.TODO(),bootstrapapi.ConfigMapClusterInfo, metav1.GetOptions{})
```
Comment (sttts) [inline cmd/kubeadm/app/discovery/file/file.go]: gofmt

File: cmd/kubeadm/app/phases/addons/dns/dns.go [hunk 3]
```diff
@@ -273,16 +274,17 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
 		return fmt.Errorf("unable to decode the DNS service %v", err)
 	}
 
+	ctx := context.TODO()
```
Comment (sttts) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: to simplify the PR, can we just use `context.TODO()` directly in all calls everywhere? I.e. no local ctx. We can (should!) have follow-up add the proper context logic. But for getting it type-correct, `context.TODO()` is just fine.

File: pkg/client/clientset_generated/internalclientset/typed/admissionregistration/internalversion/initializerconfiguration.go [hunk 4]
```diff
@@ -59,88 +61,96 @@ func newInitializerConfigurations(c *AdmissionregistrationClient) *initializerCo
...
 		VersionedParams(&options, scheme.ParameterCodec).
+		Context(ctx).
 		Do().
 		Into(result)
 	return
 }
 
 // List takes label and field selectors, and returns the list of InitializerConfigurations that match those selectors.
-func (c *initializerConfigurations) List(opts v1.ListOptions) (result *admissionregistration.InitializerConfigurationList, err error) {
+func (c *initializerConfigurations) List(ctx context.Context, opts v1.ListOptions) (result *admissionregistration.InitializerConfigurationList, err error) {
```
Comment (sttts) [inline pkg/client/clientset_generated/internalclientset/typed/admissionregistration/internalversion/initializerconfiguration.go]: Can we split generated code from manual changes in separate commits?

File: pkg/volume/csi/vol_data.json [hunk 5]
```diff
@@ -0,0 +1 @@
+{"driverName":"test-driver","volumeHandle":"test-vol1"}
```
Comment (sttts) [inline pkg/volume/csi/vol_data.json]: why this change?

File: cmd/kube-controller-manager/app/core.go [hunk 6]
```diff
@@ -28,6 +28,7 @@ import (
 
 	"github.com/golang/glog"
 
+	"context"
```
Comment (sttts) [inline cmd/kube-controller-manager/app/core.go]: order

**Comment by sttts** [comment]
Looks far simpler than I expected. It needs some work splitting generated code from manual changes, and uniformely passing `context.TODO()` instead of having a local `ctx` variable at some places.

File: pkg/volume/csi/vol_data.json (hunk 5 above)
Comment (maleck13) [reply pkg/volume/csi/vol_data.json]: will remove perhaps created as part of test. Dont see it on master

File: cmd/kubeadm/app/phases/addons/dns/dns.go (hunk 3 above)
Comment (maleck13) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: this should be done now.

**Comment by PhilipGough** [comment]
@sttts I've rebased again and fixed some of the tests. Not sure on some of the others. For example the verify job has one failing test. It say to run `hack/update-openapi-spec.sh` to update the `swagger.json` as it is outdated. However running that produces the following `git diff` which looks incorrect:

```
 {
   "swagger": "2.0",
   "info": {
-    "title": "Kubernetes",
-    "version": "v1.12.0"
+   "title": "Kubernetes",
+   "version": "v1.11.0"
   },

```

**Comment by PhilipGough** [comment]
@sttts code split up now as requested, still seeing a couple of failures

File: api/openapi-spec/swagger.json [hunk 7]
```diff
@@ -2,7 +2,7 @@
   "swagger": "2.0",
   "info": {
    "title": "Kubernetes",
-   "version": "v1.11.0"
```
Comment (sttts) [inline api/openapi-spec/swagger.json]: squash this commit into the "generated" commit.

**Comment by sttts** [comment]
Looks good. I would put the code-generator changes at the top, then generated files, then the TODO() fixup.

**Comment by PhilipGough** [comment]
> I would put the code-generator changes at the top, then generated files, then the TODO() fixup.

Done, rebased also and created the issue we spoke about with regards to generating openapi-spec

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/fake/generator_fake_for_clientset.go [hunk 8]
```diff
@@ -77,6 +77,7 @@ func (g *genClientset) Imports(c *generator.Context) (imports []string) {
 		"fakediscovery \"k8s.io/client-go/discovery/fake\"",
 		"k8s.io/apimachinery/pkg/runtime",
 		"k8s.io/apimachinery/pkg/watch",
+		"context",
```
Comment (sttts) [inline staging/src/k8s.io/code-generator/cmd/client-gen/generators/fake/generator_fake_for_clientset.go]: is this line actually needed?

**Comment by sttts** [comment]
@deads2k @lavalamp @smarterclayton @liggitt this PR is looking good now, adding the contexts to the clients and `context.TODO()` at all call sites. Do we have consensus to go forward with it now (we decided to do so in sig-api-machinery a couple of meetings back)?

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/generic.go [hunk 9]
```diff
@@ -162,14 +163,14 @@ func (f *genericInformer) Lister() {{.cacheGenericLister|raw}} {
 var forResource = `
 // ForResource gives generic access to a shared informer of the matching type
 // TODO extend this to unknown resources with a client pool
-func (f *sharedInformerFactory) ForResource(resource {{.schemaGroupVersionResource|raw}}) (GenericInformer, error) {
+func (f *sharedInformerFactory) ForResource(ctx {{.context|raw}}, resource {{.schemaGroupVersionResource|raw}}) (GenericInformer, error) {
```
Comment (liggitt) [inline staging/src/k8s.io/code-generator/cmd/informer-gen/generators/generic.go]: I'm trying to remember, what was the purpose of a context object on a shared informer that makes repeated, long-lived calls? deadline and timeout options aren't likely to be useful

**Comment by liggitt** [comment]
> @deads2k @lavalamp @smarterclayton @liggitt this PR is looking good now, adding the contexts to the clients and context.TODO() at all call sites. Do we have consensus to go forward with it now (we decided to do so in sig-api-machinery a couple of meetings back)?

I'm in favor for the methods that map 1:1 to API calls. I'm less sure about the Informer/Lister context options... what was the reason for those?

**Comment by maleck13** [comment]
@liggitt Yes wasn't sure about those either which I mentioned way back at the start of the work. Is there a potential benefit for allowing tracing hooks to passed through and invoked on the context here? Happy to take guidance on this one. The most obvious call sites are as you mention the 1:1 mappings

**Comment by liggitt** [comment]
> @liggitt Yes wasn't sure about those either which I mentioned way back at the start of the work. Is there a potential benefit for allowing tracing hooks to passed through and invoked on the context here?

given that the documentation and common use of context objects is for request-scoped data (timeout, deadline, and tracing), I don't think we should add a single context to listers/informer constructors

**Comment by liggitt** [comment]
I'm in favor of this, other than the lister/informer bits

**Comment by maleck13** [comment]
Is there anything else needed other than the removing the ctx from the listers/informers?
Any other people who need to see this? I want to avoid constantly trying to keep this up to date

**Pull request closed without merging**